### ISSUE TITLE
Proposal: OpenAPI specification to define API endpoints

### ISSUE THREAD

**Issue by gmile** [Nebo15/ehealth.api#10]
**TL;DR**. Current proposal is a result of a bit of research around available API documentation specifications. It is proposed:

* to switch from [API Blueprint](https://apiblueprint.org) promoted by Apiary in favor of [OpenAPI specification](https://github.com/OAI/OpenAPI-Specification/blob/3.0.0-rc0/versions/3.0.md) formerly known as Swagger Restful API Documentation Specification and currently promoted by Open API Initiative.

* to move the work on API spec entirely on GitHub.

Checkout this [video](https://youtu.be/4lBMQteMd6Y) for a great overview of OpenAPI Specification.

<hr>

### Apiary & API Blueprint 

Right now we use Apiary. It has proven to be _good enough_. That said, see its list of pros and cons below:

#### Pros

1. It has a native UI editor,
1. It has built-in API console,
1. It has built-in API mocks,
1. It is a hosted solution,
1. It has one-way GitHub integration,
1. It is fairly [popular](https://stackshare.io/apiary) in the wild,
1. It relies on the API Blueprint spec, which is open-source,
1. Its API Blueprint spec is a superset of markdown, a human-friendly format,
1. Its markdown extension parts are validated by the native editor.

#### Cons

1. Markdown is hard (although possible) to parse programmatically,
1. Writing API specs in a markdown’s superset means that one actually has to write the text in plain english with little-to-no structure being enforced. It requires more mental energy and is generally error-prone,
1. Markdown is too flexible. There’s no clear structure as to how you should write a spec (except for special extension blocks, provided by API Blueprint),
1. There’s no way to see history of edits,
1. There’s no way to properly collaborate (review, comment, open issues, etc.) – it has to be done  separately in Github, Slack or Confluence,
1. There’s no way to push edits from GitHub to Apiary (there’s a work around though, involving copy-pasting entire content of the spec from GitHub to Apiary after edits on GitHub),
1. Apiary is a paid service.

#### Summing up

It appears that the strongest value of Apiary is in its native spec-editing UI (arguably a user-friendly one), the ability to have mock servers and ability to test HTTP queries from the UI. The weakest part of Apiary is it’s API Blueprint.

My biggest issue with Apiary is two-fold:

1. the API Blueprint spec is not machine-readable making it hard to work with the spec programmatically.

    Particularly I’m interested in making it possible to test the actual coded implementation against API endpoint definitions found in the spec. Ability to test the spec against actually coded implementation automatically (e.g. in the CI) will significantly shorten the communication gap between the spec designers and developers, allowing for true asynchronicity of work and much shorter feedback loops,

1. Apiary has extremely poor collaboration features.

    I’m interested in lowering the bar when collaborating on the spec by leveraging features offered by GitHub,

### OpenAPI Specification

I’m proposing an alternative to Apiary and API Blueprint, a purely pragmatic approach – **write API specs in OpenAPI format directly on GitHub**:

https://github.com/OAI/OpenAPI-Specification/blob/3.0.0-rc0/versions/3.0.md

The format is a result of Apigee and Swagger [joining forces](http://swagger.io/introducing-the-open-api-initiative/) in order to make the next generation API spec format. The spec was rebranded, updated and is currently known to be driven by [Open API Initiative](https://www.openapis.org) organization.

Below are some OpenAPI’s pros and cons:

#### Pros

1. Spec is written in a pure JSON (or YAML) format, making it possible to build automation around the spec: parse it, validate it, extract pieces of it, add GitHub push hooks, etc..
1. The room for error is much decreased. One always has a structure that needs to be filled in, no need to explain things in plain english (although, you are encouraged to fill in “[description](https://github.com/APIs-guru/openapi-directory/blob/master/APIs/bbc.com/1.0.0/swagger.yaml#L170)” field).
1. The spec make heavy use of [JSON Schemas](http://json-schema.org), meaning that it is possible to store resource definition schemas alongside the spec itself, if desired,
1. It is open-sourced and is hosted on Github, with a community of possibly hundreds of eyes on it,
1. It is promoted by Google. An good overview of OpenAPI (including historical stuff) was given during Google Cloud Next ’17 conference this march – “[API design and What’s new with OpenAPI](https://youtu.be/4lBMQteMd6Y)“,
1. It is possible to use any tool (commercial or free) for rendering, testing or mocking, if desired.

    For example, an free open-source tool called [ReDoc](https://github.com/Rebilly/ReDoc) is one of the first to support the 3rd installment of OpenAPI specification. Real-world examples of ReDoc usage include the following APIs:

      * [Docker Engine](https://docs.docker.com/engine/api/v1.25),
      * [Shopify Draft Orders](https://help.shopify.com/api/draft-orders),
      * [Discourse](https://docs.discourse.org/).

1. There are tons of real API examples to learn from, all gathered in a community driven [OpenAPI Directory](https://github.com/APIs-guru/openapi-directory). Though it must be noted that most of examples are currently based on the version 2.0 of the OpenAPI spec.
1. It is [backed by](https://www.linuxfoundation.org/news-media/announcements/2015/11/new-collaborative-project-extend-swagger-specification-building) Linux Foundation,
1. Previous installments of the spec were much more popular than its competitors (the screenshot was taken from the [video](https://www.youtube.com/watch?time_continue=298&v=4lBMQteMd6Y)):

    ![image](https://cloud.githubusercontent.com/assets/207112/24403511/8373ab6c-13bd-11e7-8d26-0817fb13674e.png)

#### Cons

1. The fixed structure of the spec dictates & limits what you can or cannot write about your API endpoint. The limitless flexibility of plain text english is sacrificed in favor of strictness, correctness and recognizability,

1. OpenAPI is just a spec. Out of the box it promotes no tooling for rendering, testing or mocking,

1. OpenAPI Specification 3.0 is in its Release Candidate stage, meaning it’s not final. However at the same time RC stage means no significant (aka breaking) changes are expected to surface in the spec before the final release.

#### Summing up

The strongest part of OpenAPI Specification is the fact that it’s machine readable. It’s weakest parts are the fact that there’s no officially promoted tooling around it (at least not yet), and the relative limits in how you describe the endpoints, enforced by the structure.

I’m yet to see how we can face these limits though. Also it must be noted that in case there’s a need to describe how the API must be used (e.g. authorization steps, rate limits, best practices, etc.) – it must be done separately, probably in a place called “API Guides”. API documentation should not be mixed with API guides.

<hr>

### Closing notes

I strongly believe that moving the work on the spec entirely to a plaintext realm of GitHub will have huge long-term benefits, particularly in the area of collaboration.

At the same time keeping the spec in machine-readable format is extremely important in order to save people’s time when writing implementations against the spec, or integraions with the spec's by product, the actual API.

On a somewhat much smaller scale the usefulness of writing the specs in a machine-readable format was once againt proven to be effective when working on [os.json_schemas](https://github.com/Nebo15/os.json_schemas) (a private project of Nebo15).

It is important to note that Apiary already [supports](https://blog.apiary.io/2016/01/18/We-ve-got-Swagger) Swagger Restful API Documentation Specification v2 – see "[Choosing the right format](https://help.apiary.io/api_101/choosing-format/)" article on Apiary. It is admitted by Apiary the Swagger-kind of spec is focused on the ability to parse and integrate with the spec.

This alone, however, fulfills current proposal only partially:

* mainly, Apiary lacks collaboration capabilities comparable to GitHub,
* Apiary supports only the v2 of the spec,
* Apiary remains a service that has to be payed for.

Finally, in case the proposal is greenlit, **I volunteer to migrate existing docs to the new format**.

**Comment by AndrewDryga** [comment]
We already tried few solutions and we migrated from Swagger (current OpenAPI) to Apiry format, because documentation without context means that only back-end developer can understand it (because team meeting pushed context into his head), but no other parties will be able to use this docs.

By adding context to documentation you will end up with massive JSON file that contains raw markdown data or formatted text. Just look at official example where [json string is stored in json field](https://github.com/OAI/OpenAPI-Specification/blob/master/examples/v2.0/json/api-with-examples.json). And it's just a little part of stuff you will face by using `description` field.

I believe that our current development cycle should work better, because:
1. We can store all content in one place. (Confluence doesn't store usage details, only internal implementation, that should be apart from API docs.)
2. Developers can maintain their docs and make sure that they work with real system (thanks to debugger tool that can make real requests)
3. I or analysts can add human-readable description in easy to understand and format description.
4. We actually DO have our structured data model, because we use MSON with API Blueprint. And it's usually more up-to date than all other information sources.

I also believe that API docs and specs should be on one place, because storing them in different places be harder to maintain for us, and harder for API consumers (they need to check two places, or even know that there are TWO places and not only one request methods list).

Only because of this, [Stipe](https://stripe.com/docs/api) has one of the best docs I know.

---

Also lets take look at your cons and goals:
> Markdown is hard (although possible) to parse programmatically,

There are plenty of parsers on API Blueprint site, some of them even able to convert our docs to OpenAPI format.

> Writing API specs in a markdown’s superset means that one actually has to write the text in plain english with little-to-no structure being enforced. It requires more mental energy and is generally error-prone,

API Blueprint is all about structure on top of Markdown. We have MSON for data modeling and very structured requests/responses that don't even render if you will have some errors. 

Also Apiary generates JSON Schemas to validate requests and responses, you only need to take care about marking properties types and requirements.

> Markdown is too flexible. There’s no clear structure as to how you should write a spec (except for special extension blocks, provided by API Blueprint),

I don't see any issues in our docs, everything renders to the requests/response section just because structure is valid.

> There’s no way to see history of edits,

You can see them in GitHub commits.

> There’s no way to properly collaborate (review, comment, open issues, etc.) – it has to be done separately in Github, Slack or Confluence,

Paid version has it. But no other tool solves it. We can switch from Apiary to any other self-hosted rendering tool, but we will use plenty of features and increase our operational costs.

> There’s no way to push edits from GitHub to Apiary (there’s a work around though, involving copy-pasting entire content of the spec from GitHub to Apiary after edits on GitHub),

True.

> Apiary is a paid service.

True. But hosting and maintaining open-source alternative is also not free.

---

We can find other ways to accomplish your targets:
- Utilize API Blueprint tools to run tests or validate structures, or strip all data that you don't need and make a temporary transformations to swagger format. Eventually it's easy for you to have a machine editable version of our docs.
- We can replace Apiary with any open-source rendering tool for collaboration or pay in advance for Pro features. But there are no tools that solve collaboration out of the box, they all bases on the fact that you use PR's and later you just render never version. This approach doesn't depend on formatting spec that we will choose.


P.S.: Existing docs base it more than 10 projects. You will need month to migrate everything.

**Comment by AndrewDryga** [comment]
Summing up: humans is writing docs and humans are reading it, so it should be more suitable for humans. You still have machine-parsing properties. Collaboration difficulties is not because of format, but because of tool we use.
